{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af79e5f7-b4e3-43af-993b-969bc33aa16f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# collect brain samples from schizophrenia\n",
    "# run orfanage\n",
    "# compare to normal gtex brain for example or just refseq/gencode\n",
    "# or compare both gtex and schizophrenia to refseq/gencode separately and then compare comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "86149114",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# main imports\n",
    "import os\n",
    "import sys\n",
    "import ast\n",
    "import glob\n",
    "import math\n",
    "import shutil\n",
    "import random\n",
    "import importlib\n",
    "import subprocess\n",
    "\n",
    "from itertools import product\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.ticker import (MultipleLocator, FormatStrFormatter,\n",
    "                               AutoMinorLocator)\n",
    "import matplotlib.pylab as pylab\n",
    "from curlyBrace import curlyBrace\n",
    "import matplotlib.patches as patches\n",
    "import upsetplot\n",
    "import seaborn as sns\n",
    "\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "plt.rcParams['font.size'] = 24\n",
    "%matplotlib inline\n",
    "\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "732ab7c1-a55e-4aab-8556-3c45ae7f8655",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.insert(0, \"\") # directory to where definitions.py is\n",
    "%aimport definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6cdef1c8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# paths\n",
    "\n",
    "base_dir = \"\" # base directory for the analysis\n",
    "\n",
    "orfanage_bin = \"orfanage\" # path to the orfanage binary\n",
    "gffcompare_bin = \"\" # path to the gffcompare binary\n",
    "gffread_bin = \"\" # path to the gffread binary\n",
    "igvtools_bin = \"\" # path to the igvtools binary\n",
    "sashimi_bin = \"\" # path to the sashimi.py binary (from the TieBrush package)\n",
    "tiewrap_bin = \"tiewrap.py\" # path to the tiewrap.py binary (from the TieBrush package)\n",
    "tiecov_bin = \"tiecov\" # path to the tiecov binary (from the TieBrush package)\n",
    "\n",
    "fa_fname = \"hg38.fa\" # path to the reference genome fasta file\n",
    "chess2_gtf_fname = \"chess2.gtf\" # path to the CHESS2.2 gtf file\n",
    "chess3_gtf_fname = \"chess3.0.gtf\" # path to the CHESS3.0 gtf file\n",
    "refseq_gtf_fname = \"refseq.gtf\" # path to the refseq gtf file\n",
    "gencode_gtf_fname = \"gencode.v41.gtf\" # path to the gencode gtf file\n",
    "mane_gtf_fname = \"MANE.v10.gtf\" # path to the MANE gtf file\n",
    "gtex_gtf_fname = \"brain.gtex.gtf\" # stringtie assembly of brain samples from GTEx merged with gffcompare\n",
    "\n",
    "outdir = base_dir+\"brain_gtex/\"\n",
    "if not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)\n",
    "    \n",
    "figdir = base_dir+\"figures/\"\n",
    "if not os.path.exists(figdir):\n",
    "    os.makedirs(figdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fe0df701",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# arguments\n",
    "num_threads = 30\n",
    "ilpd = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e9188df0-765c-46f3-a0a5-d2250ae8aa41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output files\n",
    "mane_clean_gtf_fname = mane_gtf_fname.rsplit(\".\",1)[0]+\".clean.gtf\"\n",
    "refseq_clean_gtf_fname = refseq_gtf_fname.rsplit(\".\",1)[0]+\".clean.gtf\"\n",
    "gencode_clean_gtf_fname = gencode_gtf_fname.rsplit(\".\",1)[0]+\".clean.gtf\"\n",
    "\n",
    "out_gtex_refseq_gtf_fname = outdir+\"gtex.refseq.gtf\"\n",
    "out_gtex_refseq_stats_fname = outdir+\"gtex.refseq.stats\"\n",
    "out_gtex_refseq_gtf_sorted_fname = out_gtex_refseq_gtf_fname.rsplit(\".\",1)[0]+\".sorted.gtf\"\n",
    "\n",
    "mane_gtex_comp_gtf_fname = outdir+\"mane_comp.gtex.gtf\"\n",
    "mane_gtex_comp_stats_fname = outdir+\"mane_comp.gtex.stats\"\n",
    "\n",
    "refseq_gtex_comp_gtf_fname = outdir+\"refseq_comp.gtex.gtf\"\n",
    "refseq_gtex_comp_stats_fname = outdir+\"refseq_comp.gtex.stats\"\n",
    "\n",
    "gencode_gtex_comp_gtf_fname = outdir+\"gencode_comp.gtex.gtf\"\n",
    "gencode_gtex_comp_stats_fname = outdir+\"gencode_comp.gtex.stats\"\n",
    "\n",
    "\n",
    "gtex2refseq_gtf_fname = outdir+\"gtex2refseq.annotated.gtf\"\n",
    "gtex2gencode_gtf_fname = outdir+\"gtex2gencode.annotated.gtf\"\n",
    "\n",
    "\n",
    "gtex_tracking_fname = base_dir+\"data/brain.gtex.tracking\"\n",
    "out_gtex_fname = outdir+\"gtex_orf_df.tsv\"\n",
    "\n",
    "sashimi_parallel_no_orf_fname = outdir+\"top_no_orfs/sashimi.parallel\"\n",
    "sashimi_parallel_orf_fname = outdir+\"top_novel_orfs/sashimi.parallel\"\n",
    "\n",
    "junctions_fname = \"Brain.def.junctions.bed\" # junctions file generated by tiecov\n",
    "cov_fname = \"Brain.def.coverage.bedgraph\" # coverage file generated by tiecov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39917044-e14d-4ffc-8af0-1693ba8b1287",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# clean annotation\n",
    "definitions.clean(mane_gtf_fname,mane_clean_gtf_fname,gffread_bin,fa_fname,True)\n",
    "definitions.clean(refseq_gtf_fname,refseq_clean_gtf_fname,gffread_bin,fa_fname,True)\n",
    "definitions.clean(gencode_gtf_fname,gencode_clean_gtf_fname,gffread_bin,fa_fname,True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8496b101-86b3-4da4-a60d-14559442cf44",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# run orfanage gtex v refseq\n",
    "\n",
    "cmd = [orfanage_bin,\n",
    "       \"--reference\",fa_fname,\n",
    "       \"--query\",gtex_gtf_fname,\n",
    "       \"--threads\",str(num_threads),\n",
    "       \"--output\",out_gtex_refseq_gtf_fname,\n",
    "       \"--stats\",out_gtex_refseq_stats_fname,\n",
    "       \"--non_aug\",\n",
    "       refseq_clean_gtf_fname]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "igv_cmd = [igvtools_bin,\"sort\",out_gtex_refseq_gtf_fname,out_gtex_refseq_gtf_sorted_fname]\n",
    "print(\" \".join(igv_cmd))\n",
    "subprocess.call(igv_cmd)\n",
    "igv_cmd = [igvtools_bin,\"index\",out_gtex_refseq_gtf_sorted_fname]\n",
    "print(\" \".join(igv_cmd))\n",
    "subprocess.call(igv_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab8611ac-cd5b-4d62-bb12-1de00a16c631",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# we can use orfanage itself to perform all comparisons and output a stats file with all data\n",
    "# from the stats file we should be able to get every pairwise comparison and chekc whether an orf fits with no modifications or not for any reference transcript.\n",
    "# if it does - we just call it known\n",
    "\n",
    "# repeat for gtex data\n",
    "# run against mane\n",
    "cmd = [orfanage_bin,\n",
    "       \"--reference\",fa_fname,\n",
    "       \"--query\",gtex_gtf_fname,\n",
    "       \"--threads\",str(num_threads),\n",
    "       \"--output\",mane_gtex_comp_gtf_fname,\n",
    "       \"--stats\",mane_gtex_comp_stats_fname,\n",
    "       \"--non_aug\",\n",
    "       mane_gtf_fname]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "# run against refseq\n",
    "cmd = [orfanage_bin,\n",
    "       \"--reference\",fa_fname,\n",
    "       \"--query\",gtex_gtf_fname,\n",
    "       \"--threads\",str(num_threads),\n",
    "       \"--output\",refseq_gtex_comp_gtf_fname,\n",
    "       \"--stats\",refseq_gtex_comp_stats_fname,\n",
    "       \"--non_aug\",\n",
    "       refseq_gtf_fname]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "# run against gencode\n",
    "cmd = [orfanage_bin,\n",
    "       \"--reference\",fa_fname,\n",
    "       \"--query\",gtex_gtf_fname,\n",
    "       \"--threads\",str(num_threads),\n",
    "       \"--output\",gencode_gtex_comp_gtf_fname,\n",
    "       \"--stats\",gencode_gtex_comp_stats_fname,\n",
    "       \"--non_aug\",\n",
    "       gencode_gtf_fname]\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "71c7e844-a450-409c-b632-f62e039dd851",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load cds chains for all sources\n",
    "mane_cds_chains = definitions.get_chains(mane_clean_gtf_fname,\"CDS\",True)\n",
    "mane_cds_chains = mane_cds_chains[mane_cds_chains[\"has_cds\"]==1].reset_index(drop=True)\n",
    "mane_cds_chains[\"seqid\"] = mane_cds_chains[\"coords\"].str.split(\":\",n=1,expand=True)[0]\n",
    "mane_cds_chains[\"chain_str\"] = mane_cds_chains.apply(lambda row: row[\"seqid\"]+row[\"strand\"]+\",\".join([str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]]) if len(row[\"chain\"])>0 else \"-\",axis=1)\n",
    "mane_cds_chain_set = set(mane_cds_chains[\"chain_str\"])\n",
    "\n",
    "refseq_cds_chains = definitions.get_chains(refseq_clean_gtf_fname,\"CDS\",True)\n",
    "refseq_cds_chains = refseq_cds_chains[refseq_cds_chains[\"has_cds\"]==1].reset_index(drop=True)\n",
    "refseq_cds_chains[\"seqid\"] = refseq_cds_chains[\"coords\"].str.split(\":\",n=1,expand=True)[0]\n",
    "refseq_cds_chains[\"chain_str\"] = refseq_cds_chains.apply(lambda row: row[\"seqid\"]+row[\"strand\"]+\",\".join([str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]]) if len(row[\"chain\"])>0 else \"-\",axis=1)\n",
    "refseq_cds_chain_set = set(refseq_cds_chains[\"chain_str\"])\n",
    "\n",
    "gencode_cds_chains = definitions.get_chains(gencode_clean_gtf_fname,\"CDS\",True)\n",
    "gencode_cds_chains = gencode_cds_chains[gencode_cds_chains[\"has_cds\"]==1].reset_index(drop=True)\n",
    "gencode_cds_chains[\"seqid\"] = gencode_cds_chains[\"coords\"].str.split(\":\",n=1,expand=True)[0]\n",
    "gencode_cds_chains[\"chain_str\"] = gencode_cds_chains.apply(lambda row: row[\"seqid\"]+row[\"strand\"]+\",\".join([str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]]) if len(row[\"chain\"])>0 else \"-\",axis=1)\n",
    "gencode_cds_chain_set = set(gencode_cds_chains[\"chain_str\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2daaa00-27a3-43f8-90c2-b8ec9ab8685a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# run gffcompare with gencode and refseq to identify which transcripts are known\n",
    "\n",
    "# same for the gtex brain data\n",
    "gffcmp_cmd = [\"gffcompare\", \"--no-merge\",\n",
    "              \"-o\", outdir+\"gtex2refseq\",\n",
    "              \"-r\", refseq_clean_gtf_fname,\n",
    "              gtex_gtf_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "\n",
    "gffcmp_cmd = [\"gffcompare\", \"--no-merge\",\n",
    "              \"-o\", outdir+\"gtex2gencode\",\n",
    "              \"-r\", gencode_clean_gtf_fname,\n",
    "              gtex_gtf_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37085fb-ac33-41b6-98be-e57f2f76b43e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tracking = dict()\n",
    "\n",
    "with open(gtex_tracking_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.rstrip().split(\"\\t\")\n",
    "\n",
    "        tid = lcs[0]\n",
    "        tracking.setdefault(tid,[[]])\n",
    "        for q in lcs[4:]:\n",
    "            if q != \"-\":\n",
    "                tracking[tid][0].append(float(q.split(\"|\")[-3]))\n",
    "                \n",
    "# convert to dataframe\n",
    "tracking_df =  pd.DataFrame.from_dict(tracking,orient=\"index\").reset_index()\n",
    "tracking_df.columns = [\"tid\",\"tpms\"]\n",
    "\n",
    "print(\"loaded tracking\")\n",
    "\n",
    "# add ORF info\n",
    "orf_df = definitions.get_chains(out_gtex_refseq_gtf_fname,\"CDS\",True)\n",
    "orf_df = orf_df.merge(tracking_df,on=\"tid\",how=\"outer\",indicator=True)\n",
    "assert len(orf_df)==len(orf_df[orf_df[\"_merge\"]==\"both\"])\n",
    "orf_df.drop(\"_merge\",axis=1,inplace=True)\n",
    "\n",
    "orf_df[\"seqid\"] = orf_df[\"coords\"].str.split(\":\",n=1,expand=True)[0]\n",
    "orf_df[[\"num_samples\",\"max_tpm\",\"min_tpm\",\"mean_tpm\",\"sd_tpm\",\"sum_tpm\"]] = orf_df.apply(lambda row: pd.Series([len(row[\"tpms\"]),\n",
    "                                                                                                                   max(row[\"tpms\"]),\n",
    "                                                                                                                   min(row[\"tpms\"]),\n",
    "                                                                                                                   np.mean(row[\"tpms\"]),\n",
    "                                                                                                                   np.std(row[\"tpms\"]),\n",
    "                                                                                                                   np.sum(row[\"tpms\"])]), axis=1)\n",
    "\n",
    "# need to run gffcompare to identify any overlapping transcripts\n",
    "# we can limit ourselves to MANE annotation only\n",
    "gffcmp_cmd = [\"gffcompare\", \"--no-merge\",\n",
    "              \"-o\", outdir+\"gffcmp\",\n",
    "              \"-r\", mane_clean_gtf_fname,\n",
    "              out_gtex_refseq_gtf_fname]\n",
    "print(\" \".join(gffcmp_cmd))\n",
    "subprocess.call(gffcmp_cmd)\n",
    "\n",
    "# then search for the ones without CDS that overlap protein codin\n",
    "code_dict = dict()\n",
    "\n",
    "with open(outdir+\"gffcmp.annotated.gtf\",\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.rstrip().split(\"\\t\")\n",
    "        if lcs[2]==\"transcript\":\n",
    "            tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "            code = lcs[8].split(\"class_code \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "            ref_tid = \"-\"\n",
    "            if \"cmp_ref\" in lcs[8]:\n",
    "                ref_tid = lcs[8].split(\"cmp_ref \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "            code_dict[tid]=[code,ref_tid]\n",
    "\n",
    "code_df = pd.DataFrame.from_dict(code_dict,orient=\"index\").reset_index()\n",
    "code_df.columns = [\"tid\",\"code\",\"mane_tid\"]\n",
    "\n",
    "# add codes to the rest of the data\n",
    "orf_df = orf_df.merge(code_df,on=\"tid\",how=\"left\")\n",
    "\n",
    "# remove everything that does not belong\n",
    "keep_codes = [\"=\",\"c\",\"k\",\"m\",\"n\",\"j\",\"e\"]\n",
    "orf_df = orf_df[orf_df[\"code\"].isin(keep_codes)].reset_index(drop=True)\n",
    "orf_df\n",
    "\n",
    "# add gene ids from mane\n",
    "mane_gid_attr = definitions.get_attribute(mane_clean_gtf_fname,\"gene_id\")\n",
    "mane_gid_attr.columns = [\"mane_tid\",\"mane_gid\"]\n",
    "orf_df = orf_df.merge(mane_gid_attr,on=\"mane_tid\",how=\"inner\")\n",
    "\n",
    "orf_df[\"chain_str\"] = orf_df.apply(lambda row: row[\"seqid\"]+row[\"strand\"]+\",\".join([str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]]) if len(row[\"chain\"])>0 else \"-\",axis=1)\n",
    "orf_df.drop(\"chain\",inplace=True,axis=1)\n",
    "\n",
    "mane_stats_df = pd.read_csv(mane_gtex_comp_stats_fname,sep=\"\\t\")\n",
    "mane_stats_df = mane_stats_df[~(mane_stats_df[\"template_id\"]==\"-\")].reset_index(drop=True)\n",
    "mane_stats_df[\"template_len\"] = mane_stats_df[\"template_len\"].astype(int)\n",
    "mane_stats_df[\"query_len\"] = mane_stats_df[\"query_len\"].astype(int)\n",
    "mane_stats_df[\"len_extra\"] = mane_stats_df[\"len_extra\"].astype(int)\n",
    "mane_stats_df[\"len_missing\"] = mane_stats_df[\"len_missing\"].astype(int)\n",
    "mane_stats_df[\"len_outframe\"] = mane_stats_df[\"len_outframe\"].astype(int)\n",
    "mane_stats_df[\"len_match\"] = mane_stats_df[\"len_match\"].astype(int)\n",
    "mane_stats_df = mane_stats_df[~(mane_stats_df[\"template_len\"]==0)&\\\n",
    "                                  (mane_stats_df[\"template_len\"]==mane_stats_df[\"query_len\"])&\\\n",
    "                                  (mane_stats_df[\"len_extra\"]==0)&\\\n",
    "                                  (mane_stats_df[\"len_missing\"]==0)&\\\n",
    "                                  (mane_stats_df[\"len_outframe\"]==0)].reset_index(drop=True)\n",
    "\n",
    "# for every query - select the longest orf\n",
    "idx = mane_stats_df.groupby(['query_id'])['len_match'].transform(max) == mane_stats_df['len_match']\n",
    "mane_stats_df = mane_stats_df[idx].drop_duplicates([\"query_id\"],keep=\"first\")\n",
    "\n",
    "fits_orf_mane_tids = set(mane_stats_df[\"query_id\"])\n",
    "\n",
    "# add chain_str to the selected stats\n",
    "mane_orf_df = definitions.get_chains(mane_clean_gtf_fname,\"CDS\",True)\n",
    "mane_orf_df[\"chain_str\"] = mane_orf_df.apply(lambda row: row[\"seqid\"]+row[\"strand\"]+\",\".join([str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]]) if len(row[\"chain\"])>0 else \"-\",axis=1)\n",
    "\n",
    "mane_stats_df = mane_stats_df.merge(mane_orf_df[[\"tid\",\"chain_str\"]],left_on=\"template_id\",right_on=\"tid\",how=\"left\")\n",
    "mane_stats_df = mane_stats_df[[\"query_id\",\"chain_str\"]].reset_index(drop=True)\n",
    "mane_stats_df.columns = [\"tid\",\"chain_str_ref\"]\n",
    "\n",
    "# add template_id from fitting to the orf df\n",
    "orf_df = orf_df.merge(mane_stats_df,on=\"tid\",how=\"left\",indicator=False)\n",
    "# lastly replace current values with new chain_str\n",
    "orf_df[\"chain_str\"] = np.where((~orf_df[\"chain_str_ref\"].isna()) &\\\n",
    "                               (orf_df[\"tid\"].isin(fits_orf_mane_tids)),orf_df[\"chain_str_ref\"],orf_df[\"chain_str\"])\n",
    "orf_df.drop(\"chain_str_ref\",inplace=True,axis=1)\n",
    "\n",
    "# repeat for RefSeq\n",
    "\n",
    "refseq_stats_df = pd.read_csv(refseq_gtex_comp_stats_fname,sep=\"\\t\")\n",
    "refseq_stats_df = refseq_stats_df[~(refseq_stats_df[\"template_id\"]==\"-\")].reset_index(drop=True)\n",
    "refseq_stats_df[\"template_len\"] = refseq_stats_df[\"template_len\"].astype(int)\n",
    "refseq_stats_df[\"query_len\"] = refseq_stats_df[\"query_len\"].astype(int)\n",
    "refseq_stats_df[\"len_extra\"] = refseq_stats_df[\"len_extra\"].astype(int)\n",
    "refseq_stats_df[\"len_missing\"] = refseq_stats_df[\"len_missing\"].astype(int)\n",
    "refseq_stats_df[\"len_outframe\"] = refseq_stats_df[\"len_outframe\"].astype(int)\n",
    "refseq_stats_df[\"len_match\"] = refseq_stats_df[\"len_match\"].astype(int)\n",
    "refseq_stats_df = refseq_stats_df[~(refseq_stats_df[\"template_len\"]==0)&\\\n",
    "                                  (refseq_stats_df[\"template_len\"]==refseq_stats_df[\"query_len\"])&\\\n",
    "                                  (refseq_stats_df[\"len_extra\"]==0)&\\\n",
    "                                  (refseq_stats_df[\"len_missing\"]==0)&\\\n",
    "                                  (refseq_stats_df[\"len_outframe\"]==0)].reset_index(drop=True)\n",
    "\n",
    "# for every query - select the longest orf\n",
    "idx = refseq_stats_df.groupby(['query_id'])['len_match'].transform(max) == refseq_stats_df['len_match']\n",
    "refseq_stats_df = refseq_stats_df[idx].drop_duplicates([\"query_id\"],keep=\"first\")\n",
    "\n",
    "fits_orf_refseq_tids = set(refseq_stats_df[\"query_id\"])\n",
    "\n",
    "# add chain_str to the selected stats\n",
    "refseq_orf_df = definitions.get_chains(refseq_clean_gtf_fname,\"CDS\",True)\n",
    "refseq_orf_df[\"chain_str\"] = refseq_orf_df.apply(lambda row: row[\"seqid\"]+row[\"strand\"]+\",\".join([str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]]) if len(row[\"chain\"])>0 else \"-\",axis=1)\n",
    "\n",
    "refseq_stats_df = refseq_stats_df.merge(refseq_orf_df[[\"tid\",\"chain_str\"]],left_on=\"template_id\",right_on=\"tid\",how=\"left\")\n",
    "refseq_stats_df = refseq_stats_df[[\"query_id\",\"chain_str\"]].reset_index(drop=True)\n",
    "refseq_stats_df.columns = [\"tid\",\"chain_str_ref\"]\n",
    "\n",
    "# add template_id from fitting to the orf df\n",
    "orf_df = orf_df.merge(refseq_stats_df,on=\"tid\",how=\"left\",indicator=False)\n",
    "# lastly replace current values with new chain_str\n",
    "orf_df[\"chain_str\"] = np.where((~orf_df[\"chain_str_ref\"].isna()) &\\\n",
    "                               (orf_df[\"tid\"].isin(fits_orf_refseq_tids)) &\\\n",
    "                               (~orf_df[\"tid\"].isin(fits_orf_mane_tids)),orf_df[\"chain_str_ref\"],orf_df[\"chain_str\"])\n",
    "orf_df.drop(\"chain_str_ref\",inplace=True,axis=1)\n",
    "\n",
    "# repeat for gencode\n",
    "gencode_stats_df = pd.read_csv(gencode_gtex_comp_stats_fname,sep=\"\\t\")\n",
    "gencode_stats_df = gencode_stats_df[~(gencode_stats_df[\"template_id\"]==\"-\")].reset_index(drop=True)\n",
    "gencode_stats_df[\"template_len\"] = gencode_stats_df[\"template_len\"].astype(int)\n",
    "gencode_stats_df[\"query_len\"] = gencode_stats_df[\"query_len\"].astype(int)\n",
    "gencode_stats_df[\"len_extra\"] = gencode_stats_df[\"len_extra\"].astype(int)\n",
    "gencode_stats_df[\"len_missing\"] = gencode_stats_df[\"len_missing\"].astype(int)\n",
    "gencode_stats_df[\"len_outframe\"] = gencode_stats_df[\"len_outframe\"].astype(int)\n",
    "gencode_stats_df[\"len_match\"] = gencode_stats_df[\"len_match\"].astype(int)\n",
    "gencode_stats_df = gencode_stats_df[~(gencode_stats_df[\"template_len\"]==0)&\\\n",
    "                                  (gencode_stats_df[\"template_len\"]==gencode_stats_df[\"query_len\"])&\\\n",
    "                                  (gencode_stats_df[\"len_extra\"]==0)&\\\n",
    "                                  (gencode_stats_df[\"len_missing\"]==0)&\\\n",
    "                                  (gencode_stats_df[\"len_outframe\"]==0)].reset_index(drop=True)\n",
    "\n",
    "# for every query - select the longest orf\n",
    "idx = gencode_stats_df.groupby(['query_id'])['len_match'].transform(max) == gencode_stats_df['len_match']\n",
    "gencode_stats_df = gencode_stats_df[idx].drop_duplicates([\"query_id\"],keep=\"first\")\n",
    "\n",
    "fits_orf_gencode_tids = set(gencode_stats_df[\"query_id\"])\n",
    "\n",
    "# add chain_str to the selected stats\n",
    "gencode_orf_df = definitions.get_chains(gencode_clean_gtf_fname,\"CDS\",True)\n",
    "gencode_orf_df[\"chain_str\"] = gencode_orf_df.apply(lambda row: row[\"seqid\"]+row[\"strand\"]+\",\".join([str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]]) if len(row[\"chain\"])>0 else \"-\",axis=1)\n",
    "\n",
    "gencode_stats_df = gencode_stats_df.merge(gencode_orf_df[[\"tid\",\"chain_str\"]],left_on=\"template_id\",right_on=\"tid\",how=\"left\")\n",
    "gencode_stats_df = gencode_stats_df[[\"query_id\",\"chain_str\"]].reset_index(drop=True)\n",
    "gencode_stats_df.columns = [\"tid\",\"chain_str_ref\"]\n",
    "\n",
    "# add template_id from fitting to the orf df\n",
    "orf_df = orf_df.merge(gencode_stats_df,on=\"tid\",how=\"left\",indicator=False)\n",
    "# lastly replace current values with new chain_str\n",
    "orf_df[\"chain_str\"] = np.where((~orf_df[\"chain_str_ref\"].isna()) &\\\n",
    "                               (orf_df[\"tid\"].isin(fits_orf_gencode_tids)) &\\\n",
    "                               (~orf_df[\"tid\"].isin(fits_orf_refseq_tids)) &\\\n",
    "                               (~orf_df[\"tid\"].isin(fits_orf_mane_tids)),orf_df[\"chain_str_ref\"],orf_df[\"chain_str\"])\n",
    "orf_df.drop(\"chain_str_ref\",inplace=True,axis=1)\n",
    "\n",
    "orf_df[\"fits_orf_mane\"] = np.where(orf_df[\"tid\"].isin(fits_orf_mane_tids),1,0)\n",
    "orf_df[\"fits_orf_refseq\"] = np.where(orf_df[\"tid\"].isin(fits_orf_refseq_tids),1,0)\n",
    "orf_df[\"fits_orf_gencode\"] = np.where(orf_df[\"tid\"].isin(fits_orf_gencode_tids),1,0)\n",
    "\n",
    "# add known/novel flag based on whether observed in any datasets\n",
    "orf_df[\"known\"] = np.where((orf_df[\"code\"]==\"=\") |\\\n",
    "                           (orf_df[\"fits_orf_mane\"]==1) |\\\n",
    "                           (orf_df[\"fits_orf_refseq\"]==1) |\\\n",
    "                           (orf_df[\"fits_orf_gencode\"]==1),1,0)\n",
    "\n",
    "# now need to re-label non-coding reference transcripts - they should not have orf attached\n",
    "gtex2gencode_code_dict = definitions.get_attribute(gtex2gencode_gtf_fname,\"class_code\")\n",
    "gtex2gencode_match_df = gtex2gencode_code_dict[gtex2gencode_code_dict[\"class_code\"]==\"=\"].reset_index(drop=True)\n",
    "gtex2refseq_code_dict = definitions.get_attribute(gtex2refseq_gtf_fname,\"class_code\")\n",
    "gtex2refseq_match_df = gtex2refseq_code_dict[gtex2refseq_code_dict[\"class_code\"]==\"=\"].reset_index(drop=True)\n",
    "\n",
    "orf_df[\"tx_in_gencode\"] = np.where(orf_df[\"tid\"].isin(set(gtex2gencode_match_df[\"tid\"])),1,0)\n",
    "orf_df[\"tx_in_refseq\"] = np.where(orf_df[\"tid\"].isin(set(gtex2refseq_match_df[\"tid\"])),1,0)\n",
    "\n",
    "# if not known and in gencode or refseq - set as known and remove chain_str\n",
    "orf_df[\"has_cds\"] = np.where((orf_df[\"known\"]==0) &\\\n",
    "                               ((orf_df[\"tx_in_gencode\"]==1) | (orf_df[\"tx_in_refseq\"]==1)),0,orf_df[\"has_cds\"])\n",
    "orf_df[\"chain_str\"] = np.where((orf_df[\"known\"]==0) &\\\n",
    "                               ((orf_df[\"tx_in_gencode\"]==1) | (orf_df[\"tx_in_refseq\"]==1)),\"-\",orf_df[\"chain_str\"])\n",
    "orf_df[\"known\"] = np.where((orf_df[\"known\"]==0) &\\\n",
    "                               ((orf_df[\"tx_in_gencode\"]==1) | (orf_df[\"tx_in_refseq\"]==1)),1,orf_df[\"known\"])\n",
    "\n",
    "top_expressed_orf_tids = orf_df[[\"tid\",\"sum_tpm\",\"chain_str\",\"mane_gid\"]].sort_values('sum_tpm', ascending=False).drop_duplicates([\"chain_str\",\"mane_gid\"])[[\"mane_gid\",\"chain_str\",\"tid\"]]\n",
    "orf_df = orf_df.merge(top_expressed_orf_tids,on=[\"mane_gid\",\"chain_str\"],how=\"left\")\n",
    "orf_df.rename({\"tid_x\":\"tid\",\"tid_y\":\"top_expressed_tid\"},axis=1,inplace=True)\n",
    "\n",
    "for c in [\"orf_bound\",\"orf_start\",\"orf_end\"]:\n",
    "    if c in orf_df.columns:\n",
    "        orf_df.drop(c,axis=1,inplace=True)\n",
    "\n",
    "orf_df.to_csv(out_gtex_fname,index=False,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ccda16e-c36f-4792-8c84-3bb24ba61111",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gather some data on how compatible gencode, refseq and chess2 are with mane"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f51fa69b-cf21-47a4-b0dc-237522b7f855",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>coords</th>\n",
       "      <th>tpms</th>\n",
       "      <th>num_samples</th>\n",
       "      <th>max_tpm</th>\n",
       "      <th>min_tpm</th>\n",
       "      <th>mean_tpm</th>\n",
       "      <th>sd_tpm</th>\n",
       "      <th>sum_tpm</th>\n",
       "      <th>code</th>\n",
       "      <th>mane_tid</th>\n",
       "      <th>mane_gid</th>\n",
       "      <th>chain_str</th>\n",
       "      <th>fits_orf_mane</th>\n",
       "      <th>fits_orf_refseq</th>\n",
       "      <th>fits_orf_gencode</th>\n",
       "      <th>known</th>\n",
       "      <th>tx_in_gencode</th>\n",
       "      <th>tx_in_refseq</th>\n",
       "      <th>top_expressed_tid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Brain_00050473</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1324703-1328896</td>\n",
       "      <td>[2.230656, 1.845162, 1.507523, 1.779777, 1.542...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.230436</td>\n",
       "      <td>0.801009</td>\n",
       "      <td>2.161310</td>\n",
       "      <td>1.467369</td>\n",
       "      <td>34.580967</td>\n",
       "      <td>m</td>\n",
       "      <td>ENST00000343938.9</td>\n",
       "      <td>ENSG00000224051.7</td>\n",
       "      <td>chr1+1327308-1327763</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain_00050473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Brain_00000040</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1324757-1328896</td>\n",
       "      <td>[4.000858, 3.410866, 2.563985, 1.62882, 1.5822...</td>\n",
       "      <td>1396.0</td>\n",
       "      <td>32.310345</td>\n",
       "      <td>0.044356</td>\n",
       "      <td>3.354976</td>\n",
       "      <td>2.221279</td>\n",
       "      <td>4683.545906</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000343938.9</td>\n",
       "      <td>ENSG00000224051.7</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain_00000040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Brain_00000041</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1324763-1328896</td>\n",
       "      <td>[24.805258, 15.2651, 3.867779, 10.687738, 8.49...</td>\n",
       "      <td>1439.0</td>\n",
       "      <td>161.330841</td>\n",
       "      <td>0.196583</td>\n",
       "      <td>15.477317</td>\n",
       "      <td>8.596074</td>\n",
       "      <td>22271.859466</td>\n",
       "      <td>=</td>\n",
       "      <td>ENST00000343938.9</td>\n",
       "      <td>ENSG00000224051.7</td>\n",
       "      <td>chr1+1326911-1327032,1327241-1327763</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain_00000041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Brain_00233286</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1324793-1328896</td>\n",
       "      <td>[1.013539, 0.706441]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.013539</td>\n",
       "      <td>0.706441</td>\n",
       "      <td>0.859990</td>\n",
       "      <td>0.153549</td>\n",
       "      <td>1.719980</td>\n",
       "      <td>j</td>\n",
       "      <td>ENST00000343938.9</td>\n",
       "      <td>ENSG00000224051.7</td>\n",
       "      <td>chr1+1327308-1327763</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain_00050473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Brain_00040812</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:1324794-1334761</td>\n",
       "      <td>[4.779501, 4.400114]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.779501</td>\n",
       "      <td>4.400114</td>\n",
       "      <td>4.589808</td>\n",
       "      <td>0.189693</td>\n",
       "      <td>9.179615</td>\n",
       "      <td>k</td>\n",
       "      <td>ENST00000343938.9</td>\n",
       "      <td>ENSG00000224051.7</td>\n",
       "      <td>chr1+1326911-1327032,1327241-1327763</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain_00000041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              tid  has_cds seqid strand                coords  \\\n",
       "0  Brain_00050473        1  chr1      +  chr1:1324703-1328896   \n",
       "1  Brain_00000040        0  chr1      +  chr1:1324757-1328896   \n",
       "2  Brain_00000041        1  chr1      +  chr1:1324763-1328896   \n",
       "3  Brain_00233286        1  chr1      +  chr1:1324793-1328896   \n",
       "4  Brain_00040812        1  chr1      +  chr1:1324794-1334761   \n",
       "\n",
       "                                                tpms  num_samples     max_tpm  \\\n",
       "0  [2.230656, 1.845162, 1.507523, 1.779777, 1.542...         16.0    6.230436   \n",
       "1  [4.000858, 3.410866, 2.563985, 1.62882, 1.5822...       1396.0   32.310345   \n",
       "2  [24.805258, 15.2651, 3.867779, 10.687738, 8.49...       1439.0  161.330841   \n",
       "3                               [1.013539, 0.706441]          2.0    1.013539   \n",
       "4                               [4.779501, 4.400114]          2.0    4.779501   \n",
       "\n",
       "    min_tpm   mean_tpm    sd_tpm       sum_tpm code           mane_tid  \\\n",
       "0  0.801009   2.161310  1.467369     34.580967    m  ENST00000343938.9   \n",
       "1  0.044356   3.354976  2.221279   4683.545906    j  ENST00000343938.9   \n",
       "2  0.196583  15.477317  8.596074  22271.859466    =  ENST00000343938.9   \n",
       "3  0.706441   0.859990  0.153549      1.719980    j  ENST00000343938.9   \n",
       "4  4.400114   4.589808  0.189693      9.179615    k  ENST00000343938.9   \n",
       "\n",
       "            mane_gid                             chain_str  fits_orf_mane  \\\n",
       "0  ENSG00000224051.7                  chr1+1327308-1327763              0   \n",
       "1  ENSG00000224051.7                                     -              0   \n",
       "2  ENSG00000224051.7  chr1+1326911-1327032,1327241-1327763              1   \n",
       "3  ENSG00000224051.7                  chr1+1327308-1327763              0   \n",
       "4  ENSG00000224051.7  chr1+1326911-1327032,1327241-1327763              1   \n",
       "\n",
       "   fits_orf_refseq  fits_orf_gencode  known  tx_in_gencode  tx_in_refseq  \\\n",
       "0                0                 0      0              0             0   \n",
       "1                0                 0      1              1             0   \n",
       "2                1                 1      1              1             1   \n",
       "3                0                 0      0              0             0   \n",
       "4                1                 1      1              0             0   \n",
       "\n",
       "  top_expressed_tid  \n",
       "0    Brain_00050473  \n",
       "1    Brain_00000040  \n",
       "2    Brain_00000041  \n",
       "3    Brain_00050473  \n",
       "4    Brain_00000041  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sod = pd.read_csv(out_gtex_fname,sep=\"\\t\")\n",
    "sod.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2c581e70-5263-433e-85de-253949d254c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mane_tid</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000327044.7</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            mane_tid gene_name\n",
       "0  ENST00000641515.2     OR4F5\n",
       "1  ENST00000426406.4    OR4F29\n",
       "2  ENST00000332831.5    OR4F16\n",
       "3  ENST00000616016.5    SAMD11\n",
       "4  ENST00000327044.7     NOC2L"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mane_gene_names = definitions.get_attribute(mane_gtf_fname,\"gene_name\")\n",
    "mane_gene_names.rename({\"tid\":\"mane_tid\"},axis=1,inplace=True)\n",
    "mane_gene_names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2a3c0a33-b335-4ca7-870e-3e1c405c3201",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of passing genes: 464\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mane_tid</th>\n",
       "      <th>known</th>\n",
       "      <th>chain_str</th>\n",
       "      <th>sum_tpm</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>all</th>\n",
       "      <th>perc</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>novel_pass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000491306.6</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2+216499339-216499398,216499949-216500035</td>\n",
       "      <td>1.115900e+06</td>\n",
       "      <td>1.587851e+06</td>\n",
       "      <td>x</td>\n",
       "      <td>0.702774</td>\n",
       "      <td>RPL37A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000491306.6</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2+216499339-216499398,216499949-216500035</td>\n",
       "      <td>1.115900e+06</td>\n",
       "      <td>1.587851e+06</td>\n",
       "      <td>x</td>\n",
       "      <td>0.702774</td>\n",
       "      <td>RPL37A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000314138.11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr11+8684746-8684901</td>\n",
       "      <td>6.354521e+05</td>\n",
       "      <td>8.446588e+05</td>\n",
       "      <td>x</td>\n",
       "      <td>0.752318</td>\n",
       "      <td>RPL27A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000314138.11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr11+8684746-8684901</td>\n",
       "      <td>6.354521e+05</td>\n",
       "      <td>8.446588e+05</td>\n",
       "      <td>x</td>\n",
       "      <td>0.752318</td>\n",
       "      <td>RPL27A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000246802.10</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19+47756579-47756604,47756688-47756744,4775...</td>\n",
       "      <td>2.852148e+05</td>\n",
       "      <td>4.824624e+05</td>\n",
       "      <td>x</td>\n",
       "      <td>0.591165</td>\n",
       "      <td>NOP53</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8209</th>\n",
       "      <td>ENST00000335510.7</td>\n",
       "      <td>0</td>\n",
       "      <td>chr8+7488048-7488177,7489427-7489551</td>\n",
       "      <td>4.484470e-01</td>\n",
       "      <td>4.484470e-01</td>\n",
       "      <td>x</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>DEFB105B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8210</th>\n",
       "      <td>ENST00000637493.2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr9-38694776-38694826</td>\n",
       "      <td>3.813910e-01</td>\n",
       "      <td>3.813910e-01</td>\n",
       "      <td>x</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>FAM240B</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8211</th>\n",
       "      <td>ENST00000648244.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr6+52186432-52186458,52187004-52187057</td>\n",
       "      <td>1.108220e-01</td>\n",
       "      <td>1.311190e-01</td>\n",
       "      <td>x</td>\n",
       "      <td>0.845202</td>\n",
       "      <td>IL17A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8212</th>\n",
       "      <td>ENST00000543880.7</td>\n",
       "      <td>0</td>\n",
       "      <td>chr9-24543733-24543754,24544227-24544237</td>\n",
       "      <td>1.266660e-01</td>\n",
       "      <td>1.266660e-01</td>\n",
       "      <td>x</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>IZUMO3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8213</th>\n",
       "      <td>ENST00000340827.4</td>\n",
       "      <td>0</td>\n",
       "      <td>chr15+44973910-44973955,44978489-44978628</td>\n",
       "      <td>1.108220e-01</td>\n",
       "      <td>1.108220e-01</td>\n",
       "      <td>x</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>TERB2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8214 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                mane_tid  known  \\\n",
       "0      ENST00000491306.6      0   \n",
       "1      ENST00000491306.6      0   \n",
       "2     ENST00000314138.11      0   \n",
       "3     ENST00000314138.11      0   \n",
       "4     ENST00000246802.10      0   \n",
       "...                  ...    ...   \n",
       "8209   ENST00000335510.7      0   \n",
       "8210   ENST00000637493.2      0   \n",
       "8211   ENST00000648244.1      0   \n",
       "8212   ENST00000543880.7      0   \n",
       "8213   ENST00000340827.4      0   \n",
       "\n",
       "                                              chain_str       sum_tpm  \\\n",
       "0          chr2+216499339-216499398,216499949-216500035  1.115900e+06   \n",
       "1          chr2+216499339-216499398,216499949-216500035  1.115900e+06   \n",
       "2                                 chr11+8684746-8684901  6.354521e+05   \n",
       "3                                 chr11+8684746-8684901  6.354521e+05   \n",
       "4     chr19+47756579-47756604,47756688-47756744,4775...  2.852148e+05   \n",
       "...                                                 ...           ...   \n",
       "8209               chr8+7488048-7488177,7489427-7489551  4.484470e-01   \n",
       "8210                             chr9-38694776-38694826  3.813910e-01   \n",
       "8211           chr6+52186432-52186458,52187004-52187057  1.108220e-01   \n",
       "8212           chr9-24543733-24543754,24544227-24544237  1.266660e-01   \n",
       "8213          chr15+44973910-44973955,44978489-44978628  1.108220e-01   \n",
       "\n",
       "          gene_tpm all      perc gene_name  novel_pass  \n",
       "0     1.587851e+06   x  0.702774    RPL37A           0  \n",
       "1     1.587851e+06   x  0.702774    RPL37A           0  \n",
       "2     8.446588e+05   x  0.752318    RPL27A           0  \n",
       "3     8.446588e+05   x  0.752318    RPL27A           0  \n",
       "4     4.824624e+05   x  0.591165     NOP53           0  \n",
       "...            ...  ..       ...       ...         ...  \n",
       "8209  4.484470e-01   x  1.000000  DEFB105B           1  \n",
       "8210  3.813910e-01   x  1.000000   FAM240B           0  \n",
       "8211  1.311190e-01   x  0.845202     IL17A           0  \n",
       "8212  1.266660e-01   x  1.000000    IZUMO3           0  \n",
       "8213  1.108220e-01   x  1.000000     TERB2           0  \n",
       "\n",
       "[8214 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# contains groups of known vs novel by CDS chain\n",
    "gsod = sod[sod[\"has_cds\"]==1][[\"mane_tid\",\"known\",\"chain_str\",\"sum_tpm\"]].groupby(by=[\"mane_tid\",\"known\",\"chain_str\"]).sum().reset_index()\n",
    "gsod = gsod.merge(sod[[\"mane_tid\",\"sum_tpm\"]].groupby(by=\"mane_tid\").sum().reset_index(),on=\"mane_tid\",how=\"left\")\n",
    "gsod.columns = [\"mane_tid\",\"known\",\"chain_str\",\"sum_tpm\",\"gene_tpm\"]\n",
    "gsod[\"all\"] = \"x\"\n",
    "gsod[\"perc\"] = gsod[\"sum_tpm\"]/gsod[\"gene_tpm\"]\n",
    "gsod.sort_values(by=[\"perc\",\"gene_tpm\"],ascending=False,inplace=True)\n",
    "gsod = gsod.merge(mane_gene_names, on=\"mane_tid\",how=\"left\")\n",
    "gsod.head()\n",
    "\n",
    "# now compare chain to MANE to be able to select by ilpd\n",
    "fgsod = gsod[(gsod[\"known\"]==0)&(gsod[\"perc\"].round(2)>=0.5)].sort_values(by=\"gene_tpm\",ascending=False).reset_index(drop=True)\n",
    "print(\"number of passing genes: \"+str(len(set(fgsod[\"mane_tid\"]))))\n",
    "fsod = sod[(sod[\"has_cds\"]==1)&\\\n",
    "           (sod[\"mane_tid\"].isin(fgsod[\"mane_tid\"]))].reset_index(drop=True)\n",
    "\n",
    "# extract chain\n",
    "fsod[\"chain_orf\"] = fsod.apply(lambda row: [[int(x.split(\"-\")[-2].split(\"+\")[-1]),int(x.split(\"-\")[-1])] for x in row[\"chain_str\"].split(\",\")],axis=1)\n",
    "\n",
    "# attach mane chains\n",
    "fsod = fsod.merge(mane_cds_chains[[\"tid\",\"chain\"]].rename({\"tid\":\"mane_tid\",\"chain\":\"chain_mane\"},axis=1),on=\"mane_tid\",how=\"left\")\n",
    "\n",
    "# compare chains\n",
    "fsod[[\"mod_chain\",\n",
    "      \"c1len\",\n",
    "         \"c2len\",\n",
    "         \"match_start\",\n",
    "         \"match_end\",\n",
    "         \"num_bp_extra\",\n",
    "         \"num_bp_missing\",\n",
    "         \"num_bp_inframe\",\n",
    "         \"num_bp_match\",\n",
    "         \"num_bp_outframe\",\n",
    "         \"lpd\",\n",
    "         \"ilpd\",\n",
    "         \"mlpd\"]] = fsod.apply(lambda row: definitions.compare_and_extract(row[\"chain_orf\"],row[\"chain_mane\"],row[\"strand\"]),axis=1)\n",
    "fsod.drop([\"mod_chain\",\n",
    "          \"c1len\",\n",
    "             \"c2len\",\n",
    "             \"match_start\",\n",
    "             \"match_end\",\n",
    "             \"num_bp_extra\",\n",
    "             \"num_bp_missing\",\n",
    "             \"num_bp_inframe\",\n",
    "             \"num_bp_match\",\n",
    "             \"num_bp_outframe\",\n",
    "             \"mlpd\"],axis=1,inplace=True)\n",
    "\n",
    "# label those that pass thresholds\n",
    "fsod[\"novel_pass\"] = np.where((fsod[\"known\"]==0)&\\\n",
    "                              (fsod[\"lpd\"]>=70),1,0)\n",
    "\n",
    "# attach novel_pass to the gsod for additional selection\n",
    "fgsod = fgsod.merge(fsod[[\"chain_str\",\"novel_pass\"]],on=\"chain_str\",how=\"left\")\n",
    "fgsod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "27b74f51-545e-4a41-bc6f-753426925d79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mane_tid</th>\n",
       "      <th>known</th>\n",
       "      <th>chain_str</th>\n",
       "      <th>sum_tpm</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>all</th>\n",
       "      <th>perc</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>novel_pass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>ENST00000683415.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr15-34527632-34527834,34527922-34528077,3452...</td>\n",
       "      <td>56514.042934</td>\n",
       "      <td>105755.351968</td>\n",
       "      <td>x</td>\n",
       "      <td>0.534385</td>\n",
       "      <td>GOLGA8B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055</th>\n",
       "      <td>ENST00000265748.7</td>\n",
       "      <td>0</td>\n",
       "      <td>chr7+36406213-36406566,36407734-36407956,36410...</td>\n",
       "      <td>53287.095303</td>\n",
       "      <td>70084.240298</td>\n",
       "      <td>x</td>\n",
       "      <td>0.760329</td>\n",
       "      <td>ANLN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1320</th>\n",
       "      <td>ENST00000248846.10</td>\n",
       "      <td>0</td>\n",
       "      <td>chr22-50220045-50221874,50222028-50222102,5022...</td>\n",
       "      <td>36492.044190</td>\n",
       "      <td>60672.158806</td>\n",
       "      <td>x</td>\n",
       "      <td>0.601463</td>\n",
       "      <td>TUBGCP6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1436</th>\n",
       "      <td>ENST00000368592.8</td>\n",
       "      <td>0</td>\n",
       "      <td>chr10-132784916-132785170,132785280-132785452,...</td>\n",
       "      <td>26121.026228</td>\n",
       "      <td>51226.699324</td>\n",
       "      <td>x</td>\n",
       "      <td>0.509910</td>\n",
       "      <td>NKX6-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>ENST00000228850.6</td>\n",
       "      <td>0</td>\n",
       "      <td>chr12-4615739-4615894,4626496-4628601</td>\n",
       "      <td>44988.959469</td>\n",
       "      <td>46401.427598</td>\n",
       "      <td>x</td>\n",
       "      <td>0.969560</td>\n",
       "      <td>AKAP3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148</th>\n",
       "      <td>ENST00000266077.5</td>\n",
       "      <td>0</td>\n",
       "      <td>chr20+63741404-63741479,63741869-63742056,6374...</td>\n",
       "      <td>23379.903877</td>\n",
       "      <td>32793.545548</td>\n",
       "      <td>x</td>\n",
       "      <td>0.712942</td>\n",
       "      <td>SLC2A4RG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3241</th>\n",
       "      <td>ENST00000289734.13</td>\n",
       "      <td>0</td>\n",
       "      <td>chr8-41661466-41661564,41661876-41661941,41663...</td>\n",
       "      <td>15758.488986</td>\n",
       "      <td>31276.314988</td>\n",
       "      <td>x</td>\n",
       "      <td>0.503847</td>\n",
       "      <td>ANK1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3406</th>\n",
       "      <td>ENST00000085219.10</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19+35332873-35332924,35336036-35336341,3533...</td>\n",
       "      <td>16921.729483</td>\n",
       "      <td>24579.688740</td>\n",
       "      <td>x</td>\n",
       "      <td>0.688444</td>\n",
       "      <td>CD22</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3738</th>\n",
       "      <td>ENST00000397668.8</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19-12763282-12763431,12763528-12763599,1276...</td>\n",
       "      <td>11541.798617</td>\n",
       "      <td>20813.930859</td>\n",
       "      <td>x</td>\n",
       "      <td>0.554523</td>\n",
       "      <td>HOOK2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3878</th>\n",
       "      <td>ENST00000263094.11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19+1042662-1042826,1043041-1043251,1043334-...</td>\n",
       "      <td>10123.136959</td>\n",
       "      <td>17695.050594</td>\n",
       "      <td>x</td>\n",
       "      <td>0.572089</td>\n",
       "      <td>ABCA7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4097</th>\n",
       "      <td>ENST00000359481.9</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2+87747725-87747776,87752247-87752382,87753...</td>\n",
       "      <td>9236.768047</td>\n",
       "      <td>12096.958462</td>\n",
       "      <td>x</td>\n",
       "      <td>0.763561</td>\n",
       "      <td>PLGLB2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4654</th>\n",
       "      <td>ENST00000355705.4</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2-87016303-87016408,87017553-87017688,87022...</td>\n",
       "      <td>9788.474049</td>\n",
       "      <td>11908.438778</td>\n",
       "      <td>x</td>\n",
       "      <td>0.821978</td>\n",
       "      <td>PLGLB1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5625</th>\n",
       "      <td>ENST00000444313.8</td>\n",
       "      <td>0</td>\n",
       "      <td>chr22+44187674-44187878,44189114-44189254,4419...</td>\n",
       "      <td>4390.269285</td>\n",
       "      <td>7773.119775</td>\n",
       "      <td>x</td>\n",
       "      <td>0.564801</td>\n",
       "      <td>PARVG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5685</th>\n",
       "      <td>ENST00000559447.8</td>\n",
       "      <td>0</td>\n",
       "      <td>chr15-56093345-56096620,56098081-56098376,5610...</td>\n",
       "      <td>4099.990572</td>\n",
       "      <td>7426.429955</td>\n",
       "      <td>x</td>\n",
       "      <td>0.552081</td>\n",
       "      <td>RFX7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5692</th>\n",
       "      <td>ENST00000296043.7</td>\n",
       "      <td>0</td>\n",
       "      <td>chr4+76730877-76730935,76738761-76741926,76749...</td>\n",
       "      <td>3596.630298</td>\n",
       "      <td>6575.288039</td>\n",
       "      <td>x</td>\n",
       "      <td>0.546992</td>\n",
       "      <td>SHROOM3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6005</th>\n",
       "      <td>ENST00000329493.6</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19+3543023-3543473,3543555-3543745</td>\n",
       "      <td>2743.551764</td>\n",
       "      <td>4488.478147</td>\n",
       "      <td>x</td>\n",
       "      <td>0.611243</td>\n",
       "      <td>C19orf71</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6026</th>\n",
       "      <td>ENST00000055682.12</td>\n",
       "      <td>0</td>\n",
       "      <td>chrX-74740096-74744454</td>\n",
       "      <td>3836.894456</td>\n",
       "      <td>4433.225233</td>\n",
       "      <td>x</td>\n",
       "      <td>0.865486</td>\n",
       "      <td>NEXMIF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6055</th>\n",
       "      <td>ENST00000598519.2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19+55481206-55484373</td>\n",
       "      <td>2128.429914</td>\n",
       "      <td>4068.303065</td>\n",
       "      <td>x</td>\n",
       "      <td>0.523174</td>\n",
       "      <td>ZNF628</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6398</th>\n",
       "      <td>ENST00000272134.5</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1-225886727-225887090,225887399-225887638,2...</td>\n",
       "      <td>2140.498655</td>\n",
       "      <td>3386.131026</td>\n",
       "      <td>x</td>\n",
       "      <td>0.632137</td>\n",
       "      <td>LEFTY1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6468</th>\n",
       "      <td>ENST00000230361.4</td>\n",
       "      <td>0</td>\n",
       "      <td>chr6-42184815-42184942,42185680-42185797,42188...</td>\n",
       "      <td>2475.840837</td>\n",
       "      <td>3214.471958</td>\n",
       "      <td>x</td>\n",
       "      <td>0.770217</td>\n",
       "      <td>GUCA1B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6716</th>\n",
       "      <td>ENST00000307786.8</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1+1916492-1916615,1916752-1916896,1916974-1...</td>\n",
       "      <td>1681.945482</td>\n",
       "      <td>2317.373707</td>\n",
       "      <td>x</td>\n",
       "      <td>0.725798</td>\n",
       "      <td>CALML6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6839</th>\n",
       "      <td>ENST00000336334.8</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19-50943631-50943786,50948640-50948773,5094...</td>\n",
       "      <td>1459.212743</td>\n",
       "      <td>2176.457576</td>\n",
       "      <td>x</td>\n",
       "      <td>0.670453</td>\n",
       "      <td>KLK5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6857</th>\n",
       "      <td>ENST00000357799.9</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2+200571310-200573977,200575311-200575461,2...</td>\n",
       "      <td>1156.178008</td>\n",
       "      <td>1999.953836</td>\n",
       "      <td>x</td>\n",
       "      <td>0.578102</td>\n",
       "      <td>SGO2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6903</th>\n",
       "      <td>ENST00000419625.6</td>\n",
       "      <td>0</td>\n",
       "      <td>chr8-123681240-123681359,123684610-123684722,1...</td>\n",
       "      <td>855.412688</td>\n",
       "      <td>1502.975031</td>\n",
       "      <td>x</td>\n",
       "      <td>0.569146</td>\n",
       "      <td>ANXA13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                mane_tid  known  \\\n",
       "291    ENST00000683415.1      0   \n",
       "1055   ENST00000265748.7      0   \n",
       "1320  ENST00000248846.10      0   \n",
       "1436   ENST00000368592.8      0   \n",
       "1985   ENST00000228850.6      0   \n",
       "3148   ENST00000266077.5      0   \n",
       "3241  ENST00000289734.13      0   \n",
       "3406  ENST00000085219.10      0   \n",
       "3738   ENST00000397668.8      0   \n",
       "3878  ENST00000263094.11      0   \n",
       "4097   ENST00000359481.9      0   \n",
       "4654   ENST00000355705.4      0   \n",
       "5625   ENST00000444313.8      0   \n",
       "5685   ENST00000559447.8      0   \n",
       "5692   ENST00000296043.7      0   \n",
       "6005   ENST00000329493.6      0   \n",
       "6026  ENST00000055682.12      0   \n",
       "6055   ENST00000598519.2      0   \n",
       "6398   ENST00000272134.5      0   \n",
       "6468   ENST00000230361.4      0   \n",
       "6716   ENST00000307786.8      0   \n",
       "6839   ENST00000336334.8      0   \n",
       "6857   ENST00000357799.9      0   \n",
       "6903   ENST00000419625.6      0   \n",
       "\n",
       "                                              chain_str       sum_tpm  \\\n",
       "291   chr15-34527632-34527834,34527922-34528077,3452...  56514.042934   \n",
       "1055  chr7+36406213-36406566,36407734-36407956,36410...  53287.095303   \n",
       "1320  chr22-50220045-50221874,50222028-50222102,5022...  36492.044190   \n",
       "1436  chr10-132784916-132785170,132785280-132785452,...  26121.026228   \n",
       "1985              chr12-4615739-4615894,4626496-4628601  44988.959469   \n",
       "3148  chr20+63741404-63741479,63741869-63742056,6374...  23379.903877   \n",
       "3241  chr8-41661466-41661564,41661876-41661941,41663...  15758.488986   \n",
       "3406  chr19+35332873-35332924,35336036-35336341,3533...  16921.729483   \n",
       "3738  chr19-12763282-12763431,12763528-12763599,1276...  11541.798617   \n",
       "3878  chr19+1042662-1042826,1043041-1043251,1043334-...  10123.136959   \n",
       "4097  chr2+87747725-87747776,87752247-87752382,87753...   9236.768047   \n",
       "4654  chr2-87016303-87016408,87017553-87017688,87022...   9788.474049   \n",
       "5625  chr22+44187674-44187878,44189114-44189254,4419...   4390.269285   \n",
       "5685  chr15-56093345-56096620,56098081-56098376,5610...   4099.990572   \n",
       "5692  chr4+76730877-76730935,76738761-76741926,76749...   3596.630298   \n",
       "6005              chr19+3543023-3543473,3543555-3543745   2743.551764   \n",
       "6026                             chrX-74740096-74744454   3836.894456   \n",
       "6055                            chr19+55481206-55484373   2128.429914   \n",
       "6398  chr1-225886727-225887090,225887399-225887638,2...   2140.498655   \n",
       "6468  chr6-42184815-42184942,42185680-42185797,42188...   2475.840837   \n",
       "6716  chr1+1916492-1916615,1916752-1916896,1916974-1...   1681.945482   \n",
       "6839  chr19-50943631-50943786,50948640-50948773,5094...   1459.212743   \n",
       "6857  chr2+200571310-200573977,200575311-200575461,2...   1156.178008   \n",
       "6903  chr8-123681240-123681359,123684610-123684722,1...    855.412688   \n",
       "\n",
       "           gene_tpm all      perc gene_name  novel_pass  \n",
       "291   105755.351968   x  0.534385   GOLGA8B           1  \n",
       "1055   70084.240298   x  0.760329      ANLN           1  \n",
       "1320   60672.158806   x  0.601463   TUBGCP6           1  \n",
       "1436   51226.699324   x  0.509910    NKX6-2           1  \n",
       "1985   46401.427598   x  0.969560     AKAP3           1  \n",
       "3148   32793.545548   x  0.712942  SLC2A4RG           1  \n",
       "3241   31276.314988   x  0.503847      ANK1           1  \n",
       "3406   24579.688740   x  0.688444      CD22           1  \n",
       "3738   20813.930859   x  0.554523     HOOK2           1  \n",
       "3878   17695.050594   x  0.572089     ABCA7           1  \n",
       "4097   12096.958462   x  0.763561    PLGLB2           1  \n",
       "4654   11908.438778   x  0.821978    PLGLB1           1  \n",
       "5625    7773.119775   x  0.564801     PARVG           1  \n",
       "5685    7426.429955   x  0.552081      RFX7           1  \n",
       "5692    6575.288039   x  0.546992   SHROOM3           1  \n",
       "6005    4488.478147   x  0.611243  C19orf71           1  \n",
       "6026    4433.225233   x  0.865486    NEXMIF           1  \n",
       "6055    4068.303065   x  0.523174    ZNF628           1  \n",
       "6398    3386.131026   x  0.632137    LEFTY1           1  \n",
       "6468    3214.471958   x  0.770217    GUCA1B           1  \n",
       "6716    2317.373707   x  0.725798    CALML6           1  \n",
       "6839    2176.457576   x  0.670453      KLK5           1  \n",
       "6857    1999.953836   x  0.578102      SGO2           1  \n",
       "6903    1502.975031   x  0.569146    ANXA13           1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fgsod = fgsod[(fgsod[\"novel_pass\"]==1)&\\\n",
    "              (fgsod[\"gene_tpm\"]>=1000)].drop_duplicates()\n",
    "fgsod.reset_index(drop=True)\n",
    "fgsod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4e757f61-c75a-46b6-ae57-29fa7d6673d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sod_sub = sod[(sod[\"has_cds\"]==1)&\\\n",
    "              (sod[\"mane_tid\"].isin(fgsod[\"mane_tid\"]))].reset_index(drop=True)\n",
    "\n",
    "gene_names = dict()\n",
    "\n",
    "mane_lines = dict()\n",
    "fgsod_mane_tids = set(fgsod[\"mane_tid\"])\n",
    "with open(mane_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.split(\"\\t\")\n",
    "        if not len(lcs) == 9:\n",
    "            continue\n",
    "        tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "        if tid in fgsod_mane_tids:\n",
    "            gene_name = lcs[8].split(\"gene_name \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "            gene_names[tid] = gene_name\n",
    "            \n",
    "            mane_lines.setdefault(tid,\"\")\n",
    "            mane_lines[tid]+=line\n",
    "\n",
    "known_lines = dict()\n",
    "sod_known = sod_sub[sod_sub[\"known\"]==1][[\"top_expressed_tid\",\"sum_tpm\"]].groupby(by=\"top_expressed_tid\").sum().reset_index()\n",
    "sod_known_tpms = pd.Series(sod_known.sum_tpm.values,index=sod_known.top_expressed_tid).to_dict()\n",
    "with open(out_gtex_refseq_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.split(\"\\t\")\n",
    "        if not len(lcs) == 9:\n",
    "            continue\n",
    "        tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "        if tid in sod_known_tpms:\n",
    "            sum_tpm = round(sod_known_tpms[tid])\n",
    "            newline = \"\\t\".join(lcs[:-1])+\"\\t\"+\"transcript_id \\\"Known: \"+tid+\" (\"+str(sum_tpm)+\" TPM)\\\";\\n\"\n",
    "            known_lines.setdefault(tid,\"\")\n",
    "            known_lines[tid]+=newline\n",
    "                    \n",
    "novel_lines = dict()\n",
    "sod_novel = sod_sub[(sod_sub[\"known\"]==0)][[\"top_expressed_tid\",\"sum_tpm\"]].groupby(by=\"top_expressed_tid\").sum().reset_index()\n",
    "sod_novel_tpms = pd.Series(sod_novel.sum_tpm.values,index=sod_novel.top_expressed_tid).to_dict()\n",
    "with open(out_gtex_refseq_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.split(\"\\t\")\n",
    "        if not len(lcs) == 9:\n",
    "            continue\n",
    "        tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "        if tid in sod_novel_tpms:\n",
    "            sum_tpm = round(sod_novel_tpms[tid])\n",
    "            newline = \"\\t\".join(lcs[:-1])+\"\\t\"+\"transcript_id \\\"Novel: \"+tid+\" (\"+str(sum_tpm)+\" TPM)\\\";\\n\"\n",
    "            novel_lines.setdefault(tid,\"\")\n",
    "            novel_lines[tid]+=newline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "363e2645-952c-4155-b366-838445c70c69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# for each locus of interest - get a list of transcripts to plot\n",
    "top_selected = []\n",
    "if not os.path.exists(outdir+\"top_novel_orfs/\"):\n",
    "    os.makedirs(outdir+\"top_novel_orfs/\")\n",
    "    \n",
    "spFP = open(sashimi_parallel_orf_fname,\"w+\")\n",
    "\n",
    "for idx,row in fgsod.iterrows():\n",
    "    mane_tid = row[\"mane_tid\"]\n",
    "    tmp = sod_sub[(sod_sub[\"has_cds\"]==1)&\\\n",
    "              (sod_sub[\"mane_tid\"]==mane_tid)].reset_index(drop=True)\n",
    "    \n",
    "    if len(tmp)==0:\n",
    "        print(mane_tid)\n",
    "        continue\n",
    "        \n",
    "    tmp_known = tmp[tmp[\"known\"]==1].reset_index(drop=True) # contains all known transcripts and orfs at the locus\n",
    "    tmp_novel = tmp[(tmp[\"known\"]==0) &\\\n",
    "                    (tmp[\"chain_str\"]==row[\"chain_str\"])].reset_index(drop=True) #contains the novel transcripts with high expression that correspond to the orf\n",
    "    \n",
    "    known_tids = set(tmp_known[\"top_expressed_tid\"])\n",
    "    novel_tids = set(tmp_novel[\"top_expressed_tid\"])\n",
    "    assert len(novel_tids)==1,\"multiple novel: \"+mane_tid+\"\\t\"+row[\"chain_str\"]\n",
    "    out_base_fname = outdir+\"top_novel_orfs/\"+str(idx)+\"_\"+gene_names[mane_tid]\n",
    "\n",
    "    out_gtf_fname = out_base_fname+\".gtf\"\n",
    "    out_svg_fname = out_base_fname+\".svg\"\n",
    "\n",
    "    with open(out_gtf_fname,\"w+\") as outFP:\n",
    "        outFP.write(mane_lines[mane_tid])\n",
    "        \n",
    "        for t in known_tids:\n",
    "            outFP.write(known_lines[t])\n",
    "            \n",
    "        for t in novel_tids:\n",
    "            outFP.write(novel_lines[t])\n",
    "\n",
    "    title_str = gene_names[mane_tid]+\" \"+str(int(row[\"gene_tpm\"]))+\"TPM \"+\"{:.2f}\".format(row[\"perc\"])\n",
    "    sashimi_cmd = [sashimi_bin,\n",
    "                   \"--compare\",mane_tid,\n",
    "                   \"--sj\",junctions_fname,\n",
    "                   \"--cov\",cov_fname,\n",
    "                   \"--title\",title_str,\n",
    "                   \"--gtf\",out_gtf_fname,\n",
    "                   \"-o\",out_svg_fname]\n",
    "    spFP.write(\" \".join(sashimi_cmd)+\"\\n\")\n",
    "    \n",
    "    assert len(novel_tids)==1,\"smth wrong with novel_tids: \"+\",\".join(list(novel_tids))\n",
    "    top_selected.append([gene_names[mane_tid],mane_tid,list(novel_tids)[0],int(row[\"gene_tpm\"]),int(row[\"perc\"]*100)])\n",
    "spFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4fa36368-d7cf-4ca6-9749-bc28d139358e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = pd.DataFrame(top_selected)\n",
    "ts.columns = [\"Gene Name\",\"MANE TX\",\"Novel TX\",\"Gene Expression Novel %\",\"Gene Expression (TPM)\"]\n",
    "ts.to_csv(outdir+\"top_selected.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f329529-710b-4818-9ac9-5812452b4cc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "definitions.subset_gtf(out_gtex_refseq_gtf_fname,outdir+\"top_selected.gtf\",set(),set(ts[\"Novel TX\"]))\n",
    "\n",
    "# extract proteins for both top and chess\n",
    "cmd = [gffread_bin,\n",
    "       \"-y\",chess3_gtf_fname.rstrip(\".gtf\")+\".aa.fa\",\n",
    "       \"-g\",fa_fname,\n",
    "       chess3_gtf_fname]\n",
    "\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "cmd = [gffread_bin,\n",
    "       \"-y\",outdir+\"top_selected.aa.fa\",\n",
    "       \"-g\",fa_fname,\n",
    "       outdir+\"top_selected.gtf\"]\n",
    "\n",
    "print(\" \".join(cmd))\n",
    "subprocess.call(cmd)\n",
    "\n",
    "# load protein fastas\n",
    "chess_aa_dict = definitions.load_fasta_dict(chess3_gtf_fname.rstrip(\".gtf\")+\".aa.fa\",True)\n",
    "top_aa_dict = definitions.load_fasta_dict(outdir+\"top_selected.aa.fa\",True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38130f91-04f1-4847-92fc-f11da244982c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# now can run the file in parallel\n",
    "parallel_cmd = \"parallel -j 30 < \"+sashimi_parallel_orf_fname\n",
    "print(parallel_cmd)\n",
    "subprocess.call(parallel_cmd,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "308bc6c0-42b4-4e32-8462-bb14602a6d99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mane_tid</th>\n",
       "      <th>known</th>\n",
       "      <th>sum_tpm</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>all</th>\n",
       "      <th>perc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27772</th>\n",
       "      <td>ENST00000446656.4</td>\n",
       "      <td>0</td>\n",
       "      <td>69786.813819</td>\n",
       "      <td>69786.813819</td>\n",
       "      <td>x</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28922</th>\n",
       "      <td>ENST00000502986.4</td>\n",
       "      <td>0</td>\n",
       "      <td>56885.488307</td>\n",
       "      <td>56885.488307</td>\n",
       "      <td>x</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14235</th>\n",
       "      <td>ENST00000332836.7</td>\n",
       "      <td>0</td>\n",
       "      <td>47651.266485</td>\n",
       "      <td>47651.266485</td>\n",
       "      <td>x</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11180</th>\n",
       "      <td>ENST00000310357.8</td>\n",
       "      <td>0</td>\n",
       "      <td>35525.558227</td>\n",
       "      <td>35525.558227</td>\n",
       "      <td>x</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31045</th>\n",
       "      <td>ENST00000589042.5</td>\n",
       "      <td>0</td>\n",
       "      <td>13516.205207</td>\n",
       "      <td>13516.205207</td>\n",
       "      <td>x</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19740</th>\n",
       "      <td>ENST00000370200.6</td>\n",
       "      <td>0</td>\n",
       "      <td>632.679634</td>\n",
       "      <td>1264.471834</td>\n",
       "      <td>x</td>\n",
       "      <td>50.035091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5776</th>\n",
       "      <td>ENST00000265361.8</td>\n",
       "      <td>0</td>\n",
       "      <td>3409.140928</td>\n",
       "      <td>6815.166670</td>\n",
       "      <td>x</td>\n",
       "      <td>50.022855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7057</th>\n",
       "      <td>ENST00000278426.8</td>\n",
       "      <td>0</td>\n",
       "      <td>2794.014266</td>\n",
       "      <td>5585.976405</td>\n",
       "      <td>x</td>\n",
       "      <td>50.018369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11624</th>\n",
       "      <td>ENST00000313434.10</td>\n",
       "      <td>0</td>\n",
       "      <td>5188.767978</td>\n",
       "      <td>10374.295189</td>\n",
       "      <td>x</td>\n",
       "      <td>50.015619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29215</th>\n",
       "      <td>ENST00000514224.2</td>\n",
       "      <td>0</td>\n",
       "      <td>25205.344855</td>\n",
       "      <td>50405.734658</td>\n",
       "      <td>x</td>\n",
       "      <td>50.004915</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4190 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 mane_tid  known       sum_tpm      gene_tpm all        perc\n",
       "27772   ENST00000446656.4      0  69786.813819  69786.813819   x  100.000000\n",
       "28922   ENST00000502986.4      0  56885.488307  56885.488307   x  100.000000\n",
       "14235   ENST00000332836.7      0  47651.266485  47651.266485   x  100.000000\n",
       "11180   ENST00000310357.8      0  35525.558227  35525.558227   x  100.000000\n",
       "31045   ENST00000589042.5      0  13516.205207  13516.205207   x  100.000000\n",
       "...                   ...    ...           ...           ...  ..         ...\n",
       "19740   ENST00000370200.6      0    632.679634   1264.471834   x   50.035091\n",
       "5776    ENST00000265361.8      0   3409.140928   6815.166670   x   50.022855\n",
       "7057    ENST00000278426.8      0   2794.014266   5585.976405   x   50.018369\n",
       "11624  ENST00000313434.10      0   5188.767978  10374.295189   x   50.015619\n",
       "29215   ENST00000514224.2      0  25205.344855  50405.734658   x   50.004915\n",
       "\n",
       "[4190 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psod = sod[[\"mane_tid\",\"known\",\"sum_tpm\"]].groupby(by=[\"mane_tid\",\"known\"]).sum().reset_index()\n",
    "psod = psod.merge(sod[[\"mane_tid\",\"sum_tpm\"]].groupby(by=\"mane_tid\").sum().reset_index(),on=\"mane_tid\",how=\"left\")\n",
    "psod.columns = [\"mane_tid\",\"known\",\"sum_tpm\",\"gene_tpm\"]\n",
    "psod[\"all\"] = \"x\"\n",
    "psod[\"perc\"] = psod[\"sum_tpm\"]/psod[\"gene_tpm\"]\n",
    "psod.sort_values(by=[\"perc\",\"gene_tpm\"],ascending=False,inplace=True)\n",
    "\n",
    "psod[\"perc\"]*=100\n",
    "psod[(psod[\"known\"]==0)&(psod[\"perc\"]>=50)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4786565-1b94-4738-b261-5b3eb6b05f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.rcParams['figure.figsize'] = (12.0, 28.0)\n",
    "params = {'figure.figsize': (12,30),\n",
    "          'legend.fontsize': 50,\n",
    "          'axes.labelsize': 50,\n",
    "          'axes.titlesize': 50,\n",
    "          'xtick.labelsize': 50,\n",
    "          'ytick.labelsize': 50,\n",
    "          'font.size': 50}\n",
    "pylab.rcParams.update(params)\n",
    "\n",
    "sns.violinplot(x=\"all\",y=\"perc\",hue=\"ORF Type\",data=psod.rename({\"known\":\"ORF Type\"},axis=1).replace({\"ORF Type\":{1:\"Known\",0:\"Novel\"}}), split=True)\n",
    "f1 = plt.gcf()\n",
    "plt.ylabel(\"% Gene Expression\")\n",
    "ax = plt.gca()\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "plt.legend(loc=6)\n",
    "xlims = ax.get_xlim()\n",
    "print(xlims)\n",
    "ax.set_xlim(-0.45,0.38)\n",
    "\n",
    "axins_psod = psod[(psod[\"known\"]==0)&(psod[\"perc\"]>=50)].reset_index(drop=True)\n",
    "print(len(axins_psod))\n",
    "\n",
    "# sub region of the original image\n",
    "x1, x2, y1, y2 = -0.01, 0.08, 50, 110\n",
    "\n",
    "plt.xticks(visible=False)\n",
    "rect = patches.Rectangle((-0.01,50), 0.09, 60, linewidth=5,color=\"#B90E0A\", fc=\"none\")\n",
    "\n",
    "# Add the patch to the Axes\n",
    "ax.add_patch(rect)\n",
    "\n",
    "\n",
    "# curly bracket 1 start point and end point\n",
    "p1 = [0.08,50]\n",
    "p2 = [0.08,110]\n",
    "\n",
    "# fontdict for curly bracket 1 text\n",
    "font = {'family': 'serif',\n",
    "        'color':  'k',\n",
    "        'weight': 'bold',\n",
    "        'size': 48,\n",
    "        }\n",
    "\n",
    "# coefficient for curly bracket 1\n",
    "k_r1 = 0.05\n",
    "\n",
    "# clockwise, need to swap the start point and end point\n",
    "curlyBrace(f1, ax, p2, p1, k_r1, bool_auto=True, str_text=str(len(axins_psod)), color='black', lw=5, int_line_num=2, fontdict=font)\n",
    "\n",
    "\n",
    "plt.savefig(outdir+\"gene_expr_perc.svg\")\n",
    "print(outdir+\"gene_expr_perc.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f5b12088-4405-4e9b-b62b-ce35e50a1632",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mane_tid</th>\n",
       "      <th>known</th>\n",
       "      <th>chain_str</th>\n",
       "      <th>sum_tpm</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>all</th>\n",
       "      <th>perc</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000648839.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr17-8377564-8377692,8379796-8379936,8382143-...</td>\n",
       "      <td>1.165679e+06</td>\n",
       "      <td>1.165679e+06</td>\n",
       "      <td>x</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RPL26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000372728.4</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX-103062897-103063274</td>\n",
       "      <td>5.155847e+05</td>\n",
       "      <td>5.155847e+05</td>\n",
       "      <td>x</td>\n",
       "      <td>1.0</td>\n",
       "      <td>BEX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000314133.4</td>\n",
       "      <td>1</td>\n",
       "      <td>chr11+63974681-63974794,63976225-63976320</td>\n",
       "      <td>4.871259e+05</td>\n",
       "      <td>4.871259e+05</td>\n",
       "      <td>x</td>\n",
       "      <td>1.0</td>\n",
       "      <td>COX8A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000347512.8</td>\n",
       "      <td>1</td>\n",
       "      <td>chr19+5690508-5690600,5691319-5691453,5691532-...</td>\n",
       "      <td>4.593722e+05</td>\n",
       "      <td>4.593722e+05</td>\n",
       "      <td>x</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RPL36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000382581.5</td>\n",
       "      <td>1</td>\n",
       "      <td>chr4+6640863-6641246</td>\n",
       "      <td>4.303979e+05</td>\n",
       "      <td>4.303979e+05</td>\n",
       "      <td>x</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MRFAP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            mane_tid  known  \\\n",
       "0  ENST00000648839.1      1   \n",
       "1  ENST00000372728.4      1   \n",
       "2  ENST00000314133.4      1   \n",
       "3  ENST00000347512.8      1   \n",
       "4  ENST00000382581.5      1   \n",
       "\n",
       "                                           chain_str       sum_tpm  \\\n",
       "0  chr17-8377564-8377692,8379796-8379936,8382143-...  1.165679e+06   \n",
       "1                           chrX-103062897-103063274  5.155847e+05   \n",
       "2          chr11+63974681-63974794,63976225-63976320  4.871259e+05   \n",
       "3  chr19+5690508-5690600,5691319-5691453,5691532-...  4.593722e+05   \n",
       "4                               chr4+6640863-6641246  4.303979e+05   \n",
       "\n",
       "       gene_tpm all  perc gene_name  \n",
       "0  1.165679e+06   x   1.0     RPL26  \n",
       "1  5.155847e+05   x   1.0      BEX1  \n",
       "2  4.871259e+05   x   1.0     COX8A  \n",
       "3  4.593722e+05   x   1.0     RPL36  \n",
       "4  4.303979e+05   x   1.0    MRFAP1  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psod = sod[[\"mane_tid\",\"known\",\"sum_tpm\"]].groupby(by=[\"mane_tid\",\"known\"]).sum().reset_index()\n",
    "psod = psod.merge(sod[[\"mane_tid\",\"sum_tpm\"]].groupby(by=\"mane_tid\").sum().reset_index(),on=\"mane_tid\",how=\"left\")\n",
    "psod.columns = [\"mane_tid\",\"known\",\"sum_tpm\",\"gene_tpm\"]\n",
    "psod[\"all\"] = \"x\"\n",
    "psod[\"perc\"] = psod[\"sum_tpm\"]/psod[\"gene_tpm\"]\n",
    "psod.sort_values(by=[\"perc\",\"gene_tpm\"],ascending=False,inplace=True)\n",
    "psod.head()\n",
    "\n",
    "gsod = sod[sod[\"has_cds\"]==1][[\"mane_tid\",\"known\",\"chain_str\",\"sum_tpm\"]].groupby(by=[\"mane_tid\",\"known\",\"chain_str\"]).sum().reset_index()\n",
    "gsod = gsod.merge(sod[[\"mane_tid\",\"sum_tpm\"]].groupby(by=\"mane_tid\").sum().reset_index(),on=\"mane_tid\",how=\"left\")\n",
    "gsod.columns = [\"mane_tid\",\"known\",\"chain_str\",\"sum_tpm\",\"gene_tpm\"]\n",
    "gsod[\"all\"] = \"x\"\n",
    "gsod[\"perc\"] = gsod[\"sum_tpm\"]/gsod[\"gene_tpm\"]\n",
    "gsod.sort_values(by=[\"perc\",\"gene_tpm\"],ascending=False,inplace=True)\n",
    "gsod = gsod.merge(mane_gene_names, on=\"mane_tid\",how=\"left\")\n",
    "gsod.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4faf5838-977b-404d-ba1e-54d6dc2d54bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract gtf corresponding to the selected highly expressed novel orfs\n",
    "sub_psod = psod[(psod[\"known\"]==0)&(psod[\"perc\"]>=0.85)].reset_index(drop=True)\n",
    "sub_sod = sod[sod[\"mane_tid\"].isin(sub_psod[\"mane_tid\"])][[\"tid\",\"mane_tid\",\"has_cds\",\"num_samples\",\"sum_tpm\",\"chain_str\",\"known\"]].sort_values(by=\"sum_tpm\")\n",
    "sub_sod.reset_index(drop=True,inplace=True)\n",
    "sub_sod.head()\n",
    "\n",
    "# extract gtf\n",
    "\n",
    "sub_sod_tids = set(sub_sod[\"tid\"])\n",
    "\n",
    "with open(outdir+\"sub_sod.gtf\",\"w\") as outFP:\n",
    "    with open(out_gtex_refseq_gtf_fname,\"r\") as inFP:\n",
    "        for line in inFP:\n",
    "            lcs = line.split(\"\\t\")\n",
    "            if not len(lcs) == 9:\n",
    "                continue\n",
    "            tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "            if tid in sub_sod_tids:\n",
    "                outFP.write(line)\n",
    "\n",
    "fpsod = psod[(psod[\"known\"]==0)&(psod[\"perc\"]>=0.85)].sort_values(by=\"gene_tpm\",ascending=False).reset_index(drop=True)\n",
    "sod_sub = sod[(sod[\"has_cds\"]==1)&\\\n",
    "              (sod[\"mane_tid\"].isin(fpsod[\"mane_tid\"]))].reset_index(drop=True)\n",
    "\n",
    "gene_names = dict()\n",
    "\n",
    "mane_lines = dict()\n",
    "fpsod_mane_tids = set(fpsod[\"mane_tid\"])\n",
    "with open(mane_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.split(\"\\t\")\n",
    "        if not len(lcs) == 9:\n",
    "            continue\n",
    "        tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "        if tid in fpsod_mane_tids:\n",
    "            gene_name = lcs[8].split(\"gene_name \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "            gene_names[tid] = gene_name\n",
    "            \n",
    "            mane_lines.setdefault(tid,\"\")\n",
    "            mane_lines[tid]+=line\n",
    "\n",
    "known_lines = dict()\n",
    "sod_known = sod_sub[sod_sub[\"known\"]==1][[\"top_expressed_tid\",\"sum_tpm\"]].groupby(by=\"top_expressed_tid\").sum().reset_index()\n",
    "sod_known_tpms = pd.Series(sod_known.sum_tpm.values,index=sod_known.top_expressed_tid).to_dict()\n",
    "with open(out_gtex_refseq_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.split(\"\\t\")\n",
    "        if not len(lcs) == 9:\n",
    "            continue\n",
    "        tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "        if tid in sod_known_tpms:\n",
    "            sum_tpm = round(sod_known_tpms[tid])\n",
    "            newline = \"\\t\".join(lcs[:-1])+\"\\t\"+\"transcript_id \\\"Known: \"+tid+\" (\"+str(sum_tpm)+\" TPM)\\\";\\n\"\n",
    "            known_lines.setdefault(tid,\"\")\n",
    "            known_lines[tid]+=newline\n",
    "                    \n",
    "novel_lines = dict()\n",
    "sod_novel = sod_sub[(sod_sub[\"known\"]==0)][[\"top_expressed_tid\",\"sum_tpm\"]].groupby(by=\"top_expressed_tid\").sum().reset_index()\n",
    "sod_novel_tpms = pd.Series(sod_novel.sum_tpm.values,index=sod_novel.top_expressed_tid).to_dict()\n",
    "with open(out_gtex_refseq_gtf_fname,\"r\") as inFP:\n",
    "    for line in inFP:\n",
    "        lcs = line.split(\"\\t\")\n",
    "        if not len(lcs) == 9:\n",
    "            continue\n",
    "        tid = lcs[8].split(\"transcript_id \\\"\", 1)[1].split(\"\\\"\", 1)[0]\n",
    "        if tid in sod_novel_tpms:\n",
    "            sum_tpm = round(sod_novel_tpms[tid])\n",
    "            newline = \"\\t\".join(lcs[:-1])+\"\\t\"+\"transcript_id \\\"Novel: \"+tid+\" (\"+str(sum_tpm)+\" TPM)\\\";\\n\"\n",
    "            novel_lines.setdefault(tid,\"\")\n",
    "            novel_lines[tid]+=newline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "9bc6462f-668b-48bb-8157-2a1661c78272",
   "metadata": {},
   "outputs": [],
   "source": [
    "skipped_mane_tids = set()\n",
    "written_mane_tids = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "3afef7b6-305a-411a-b015-b5214232fd46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each locus of interest - get a list of transcripts to plot\n",
    "\n",
    "with open(outdir+\"top_with_orf.gtf\",\"w+\") as outFP:\n",
    "    for idx,row in fgsod.iterrows():\n",
    "        mane_tid = row[\"mane_tid\"]\n",
    "        written_mane_tids.add(mane_tid)\n",
    "        tmp = sod_sub[(sod_sub[\"has_cds\"]==1)&\\\n",
    "                  (sod_sub[\"mane_tid\"]==mane_tid)].reset_index(drop=True)\n",
    "\n",
    "        if len(tmp)==0:\n",
    "            skipped_mane_tids.add(mane_tid)\n",
    "            continue\n",
    "\n",
    "        tmp_known = tmp[tmp[\"known\"]==1].reset_index(drop=True) # contains all known transcripts and orfs at the locus\n",
    "        tmp_novel = tmp[(tmp[\"known\"]==0) &\\\n",
    "                        (tmp[\"chain_str\"]==row[\"chain_str\"])].reset_index(drop=True) #contains the novel transcripts with high expression that correspond to the orf\n",
    "\n",
    "        known_tids = set(tmp_known[\"top_expressed_tid\"])\n",
    "        novel_tids = set(tmp_novel[\"top_expressed_tid\"])\n",
    "        assert len(novel_tids)==1,\"multiple novel: \"+mane_tid+\"\\t\"+row[\"chain_str\"]\n",
    "\n",
    "        outFP.write(mane_lines[mane_tid])\n",
    "\n",
    "        for t in known_tids:\n",
    "            outFP.write(known_lines[t])\n",
    "\n",
    "        for t in novel_tids:\n",
    "            outFP.write(novel_lines[t])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
